File: src/features/ingestion/ingestion-contacts.controller.ts
import { Body, Controller, Post, Req, UseGuards } from '@nestjs/common';
import { Request } from 'express';
import { Roles } from '../../common/decorators/roles.decorator';
import { OrgScopeGuard } from '../../common/guards/org-scope.guard';
import { RolesGuard } from '../../common/guards/roles.guard';
import { SessionGuard } from '../../common/guards/session.guard';
import { Role } from '../../common/roles';
import { IngestionContactsEnrichDto } from './dto/ingestion-contacts-enrich.dto';
import { IngestionContactsParseRowDto } from './dto/ingestion-contacts-parse-row.dto';
import { IngestionContactsSuggestMappingDto } from './dto/ingestion-contacts-suggest-mapping.dto';
import { IngestionContactsService } from './ingestion-contacts.service';

@Controller('ingestion/contacts/v1')
@UseGuards(SessionGuard, OrgScopeGuard, RolesGuard)
export class IngestionContactsController {
  constructor(private readonly ingestion: IngestionContactsService) {}

  private actor(req: Request) {
    const user = (req as any).user || req.session?.user;
    return { userId: user?.id, orgId: user?.orgId, role: user?.role as Role | undefined };
  }

  @Roles(Role.Admin, Role.OrgOwner, Role.OrgAdmin, Role.SuperAdmin)
  @Post('suggest-mapping')
  suggestMapping(@Body() dto: IngestionContactsSuggestMappingDto, @Req() req: Request) {
    return this.ingestion.suggestMapping(this.actor(req), dto);
  }

  @Roles(Role.Admin, Role.OrgOwner, Role.OrgAdmin, Role.SuperAdmin)
  @Post('parse-row')
  parseRow(@Body() dto: IngestionContactsParseRowDto, @Req() req: Request) {
    return this.ingestion.parseRow(this.actor(req), dto);
  }

  @Roles(Role.Admin, Role.OrgOwner, Role.OrgAdmin, Role.SuperAdmin)
  @Post('enrich')
  enrich(@Body() dto: IngestionContactsEnrichDto, @Req() req: Request) {
    return this.ingestion.enrich(this.actor(req), dto);
  }
}

File: spring-boot-api/src/main/java/com/mytegroup/api/controller/ingestion/IngestionContactsController.java
package com.mytegroup.api.controller.ingestion;

import com.mytegroup.api.dto.ingestion.*;
import com.mytegroup.api.service.ingestion.IngestionContactsService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/api/ingestion/contacts/v1")
@PreAuthorize("isAuthenticated()")
@RequiredArgsConstructor
public class IngestionContactsController {

    private final IngestionContactsService ingestionContactsService;

    @PostMapping("/suggest-mapping")
    @PreAuthorize("hasAnyRole('ORG_OWNER', 'ORG_ADMIN', 'ADMIN', 'SUPER_ADMIN', 'PLATFORM_ADMIN')")
    public ResponseEntity<?> suggestMapping(
            @RequestBody @Valid IngestionContactsSuggestMappingDto dto,
            @RequestParam(required = false) String orgId) {
        
        if (orgId == null) { 
            return ResponseEntity.badRequest().body(Map.of("error", "orgId is required")); 
        }
        Map<String, Object> mapping = ingestionContactsService.suggestMapping(
            dto.getHeaders(),
            dto.getSampleRows(),
            orgId
        );
        
        return ResponseEntity.ok(mapping);
    }

    @PostMapping("/parse-row")
    @PreAuthorize("hasAnyRole('ORG_OWNER', 'ORG_ADMIN', 'ADMIN', 'SUPER_ADMIN', 'PLATFORM_ADMIN')")
    public ResponseEntity<?> parseRow(
            @RequestBody @Valid IngestionContactsParseRowDto dto,
            @RequestParam(required = false) String orgId) {
        
        if (orgId == null) { 
            return ResponseEntity.badRequest().body(Map.of("error", "orgId is required")); 
        }
        Map<String, Object> result = ingestionContactsService.parseRow(
            dto.getRow(),
            dto.getMapping(),
            orgId
        );
        
        return ResponseEntity.ok(result);
    }

    @PostMapping("/enrich")
    @PreAuthorize("hasAnyRole('ORG_OWNER', 'ORG_ADMIN', 'ADMIN', 'SUPER_ADMIN', 'PLATFORM_ADMIN')")
    public ResponseEntity<?> enrich(
            @RequestBody @Valid IngestionContactsEnrichDto dto,
            @RequestParam(required = false) String orgId) {
        
        if (orgId == null) { 
            return ResponseEntity.badRequest().body(Map.of("error", "orgId is required")); 
        }
        Map<String, Object> result = ingestionContactsService.enrich(
            dto.getContact(),
            orgId
        );
        
        return ResponseEntity.ok(result);
    }
    
        }
        
